Address Change Updates

jucarson opened the pull request:
Added ability to override conversations with an email if errors are encountered..

tkainz reviewed (commented):
mostly minor stuff but one question on the 504 response code and where your sending the error detail to the UX.

jordan-keenan reviewed (changes requested):
I think this is a good Idea Justin, and I appreciate your work to
improve the options the CU's have for this Library code.

It looks like some changes may be needed. I could be just misunderstanding how conversations is triggered in some of these cases, so please explain if I've misunderstood.

1. The way I read the Error codes, 503 means Error processing requested changes.
So in the following if condition statement, shouldn't RESPONSECODE<>504 actually
be RESPONSECODE=503?, since I don't think the other error codes would trigger conversations,
as they are memo mode, config error and ineligible account type errors?
 IF ((PARAMOVERRIDECONVS=TRUE AND
      RESPONSECODE<>504) AND
     (PARAMTESTACCTS="" OR
     (PARAMTESTACCTS<>"" AND
      TESTUSER=TRUE))) THEN
  CALL OVERRIDECONVERSATIONS
Also, there is an exception to what error code 503 means in one place in this PowerOn's code that I've pasted here;
so I don't know if you would need to consider that too (it looks like in that case a change didn't fail, just that there were no names that could be changed, but I'm not sure)?
 IF TOTALNAMECOUNT=0 THEN
  DO
   CALL JSONOPEN
   CALL PRINTPROGRAMINFO
   RESPONSECODE=503
   ERROR503CHR="No allowed name records found"
   PARSEDTEXT(1)="No allowed name records found"

   CALL ERRORHANDLER
   CALL CREATENOTERECORD
  END


2. and if it is supposed to be looking for RESPONSECODE=503,
looking at these eleven code sections,
wouldn't the note record have already been created before
the new code adds lines 3-6 to put in the note?
         IF NAMELOC=0 THEN
          DO
           RESPONSECODE=503
           PARSEDTEXT(1)=FORMAT("Banno Chg Addr Err: Name LOC:####9",NAMELOC)
           ERROR503CHR=FORMAT("Could not locate name LOC:####9",NAMELOC)
           CALL CREATENOTERECORD
          END
        END
...
         IF FMERROR<>"" THEN
          DO
           RESPONSECODE=503
           ERROR503CHR="Banno Chg Addr Err: FM-"+FMERROR
           PARSEDTEXT(1)="Banno Chg Addr Err: FM-"+FMERROR
           CALL CREATENOTERECORD
          END
...
     ELSE IF FMERROR<>"" THEN
      DO
       RESPONSECODE=503
       ERROR503CHR=FORMAT("Banno Chg Addr Err - Nm LOC #####9",ACCTLVLLOCS(TMPLOOP))+" FM-"+FMERROR
       PARSEDTEXT(1)=FORMAT("Banno Chg Addr Err - Nm LOC #####9",ACCTLVLLOCS(TMPLOOP))
       PARSEDTEXT(2)="FM-"+FMERROR
       CALL CREATENOTERECORD
      END
...
     ELSE IF FMERROR<>"" THEN
      DO
       RESPONSECODE=503
       ERROR503CHR=FORMAT("Banno Chg Addr Err - Shr Nm LOC #####9",SHARELVLLOCS(TMPLOOP))+" FM-"+FMERROR
       PARSEDTEXT(1)=FORMAT("Banno Chg Addr Err - Shr Nm LOC #####9",SHARELVLLOCS(TMPLOOP))
       PARSEDTEXT(2)="FM-"+FMERROR
       CALL CREATENOTERECORD
      END
...
     ELSE IF FMERROR<>"" THEN
      DO
       RESPONSECODE=503
       ERROR503CHR=FORMAT("Banno Chg Addr Err - IRS Nm LOC #####9",IRSLVLLOCS(TMPLOOP))+" FM-"+FMERROR
       PARSEDTEXT(1)=FORMAT("Banno Chg Addr Err - IRS Nm LOC #####9",IRSLVLLOCS(TMPLOOP))
       PARSEDTEXT(2)="FM-"+FMERROR
       CALL CREATENOTERECORD
      END
...
     ELSE IF FMERROR<>"" THEN
      DO
       RESPONSECODE=503
       ERROR503CHR=FORMAT("Banno Chg Addr Err - Ln Nm LOC #####9",LOANLVLLOCS(TMPLOOP))+" FM-"+FMERROR
       PARSEDTEXT(1)=FORMAT("Banno Chg Addr Err - Ln Nm LOC #####9",LOANLVLLOCS(TMPLOOP))
       PARSEDTEXT(2)="FM-"+FMERROR
       CALL CREATENOTERECORD
...
     ELSE IF FMERROR<>"" THEN
      DO
       RESPONSECODE=503
       ERROR503CHR=FORMAT("Banno Chg Addr Err - Pldg Nm LOC #####9",PLEDGELVLLOCS(TMPLOOP))+" FM-"+FMERROR
       PARSEDTEXT(1)=FORMAT("Banno Chg Addr Err - Pldg Nm LOC #####9",PLEDGELVLLOCS(TMPLOOP))
       PARSEDTEXT(2)="FM-"+FMERROR
       CALL CREATENOTERECORD
      END
...
     ELSE IF FMERROR<>"" THEN
      DO
       RESPONSECODE=503
       ERROR503CHR=FORMAT("Banno Chg Addr Err - EFT Nm LOC #####9",EFTLVLLOCS(TMPLOOP))+" FM-"+FMERROR
       PARSEDTEXT(1)=FORMAT("Banno Chg Addr Err - EFT Nm LOC #####9",EFTLVLLOCS(TMPLOOP))
       PARSEDTEXT(2)="FM-"+FMERROR
       CALL CREATENOTERECORD
      END
...
     ELSE IF FMERROR<>"" THEN
      DO
       RESPONSECODE=503
       ERROR503CHR=FORMAT("Banno Chg Addr Err - Card Nm LOC #####9",CARDLVLLOCS(TMPLOOP))+" FM-"+FMERROR
       PARSEDTEXT(1)=FORMAT("Banno Chg Addr Err - Card Nm LOC #####9",CARDLVLLOCS(TMPLOOP))
       PARSEDTEXT(2)="FM-"+FMERROR
       CALL CREATENOTERECORD
      END
...
     ELSE IF FMERROR<>"" THEN
      DO
       RESPONSECODE=503
       ERROR503CHR=FORMAT("Banno Chg Addr Err - X Ln Nm LOC #####9",EXTERNALLVLLOCS(TMPLOOP))+" FM-"+FMERROR
       PARSEDTEXT(1)=FORMAT("Banno Chg Addr Err - X Ln Nm LOC #####9",EXTERNALLVLLOCS(TMPLOOP))
       PARSEDTEXT(2)="FM-"+FMERROR
       CALL CREATENOTERECORD
      END
...
 IF TRACKINGFOUND=FALSE OR
   (TRACKINGFOUND=TRUE AND MEMBERADDRLINK="") THEN
  DO
   RESPONSECODE=503

   IF TRACKINGFOUND=FALSE THEN
    DO
     PARSEDTEXT(1)="Account tracking type 08 not found"
     ERROR503CHR="Tracking 8 not found"
    END
   ELSE IF (TRACKINGFOUND=TRUE AND MEMBERADDRLINK="") THEN
    DO
     PARSEDTEXT(1)="Tracking 8 found but no Mbr Addr link"
     ERROR503CHR="Tracking 8 without MbrAddr Link"
    END
   CALL CREATENOTERECORD

3. In the case where an error occurs during address change,
I think it would be a good Idea to send an email to the member
at the same time as making the note record
with similar language that Banno shows members for when an address
change has an error and has to go to conversations.
This way the member would know there was an issue,
that will be manually changed by the CU, and that would
give them the idea possibly double check their address
in Banno in a few days to make sure the cu made the manual change.
You might want to ask the BA's to see if they think it should
be part of it.
 Or another way would add a cu customizable 
message for that saturation to the cfg and contract that, 
gets a different new error code,
so the UX Team can display the custom message to the member, without sending them to conversations.

4. I think adding Debug to the Email sending would be good (like is in
the procedure SENDEMAIL). That way the logs also would have any email error
like this code in the SENDMAIL procedure:
 BNODEBUGMSG= "SEND EMAIL ERROR:"+EMAILSENDERROR
 CALL BNODEBUGMSGADD



5. Miscellaneous spelling to look at.

I know this is old code but Seperate
could be changed to Separate here:

PROCEDURE PARSEDATALINE
[* Parse through a line of data, looking for a pipe delimiter.
** Seperate text at the delimiter storing each segment into
** array elements

Same here: sucessfull can be changed to successful

PROCEDURE READCONFIGFILESETTINGS
[* Read parameter settings from the configuration Letter file.
**  IAT    PARAMINELIGIBLEACCTS  Ineligible Account Type(s) (in LISTEXPAND format)
**  AW     PARAMWARNINGSACCT     Acct level warnings which exclude account (in LISTEXPAND format)
**  CW     PARAMCLEARWARNING     Account level warning(s) to clear upon successfull address change

Same here:tto can be changed to to.

PROCEDURE POSTFMPROCESS
[* Clears and sets warning codes upon successful FM. Sends
** confirmation email tto member.

jucarson commented on change-address/LETTERSPECS/BANNO.CHANGE.ADDR.V1.CFG:
```diff
@@ -117,7 +117,7 @@ MN: ACCOUNT
 *
 * TEST ACCOUNTS 
-*  Version 1.1.0 - Testing Name Matching by level and member email.
+*  Version 1.1.1 - Testing Conversations Override changes.
```
Done

jucarson commented on change-address/LETTERSPECS/BANNO.CHANGE.ADDR.V1.CFG:
```diff
@@ -128,42 +128,62 @@ MT: 0
+OC:
+*
+*  CONVERSATIONS OVERRIDE EMAIL
+*  If OVERRIDE CONVERSATIONS above is set to true, a valid email
```
Technically it is being validated but because the email is their only notification it was invalid and they'll never get it, I added a note that indicates that with !!!'s around it.

jucarson commented on change-address/REPWRITERSPECS/BANNO.CHANGE.ADDR.V1.POW:
```diff
@@ -26,6 +26,9 @@
 **                         Updated CFG file read error msg. Set param for note expiration days to 30
 **    Ver. 1.1.0  07/27/22 J. Carson - Added old address matching functionality to override 
 **                                     tracking 8 linking. Added confirmation email to be sent to member.
+**    Ver. 1.1.1  08/10/23 J. Carson - Added option of sending email to CU and adding address information 
```
Done

jucarson commented on change-address/REPWRITERSPECS/BANNO.CHANGE.ADDR.V1.POW:
```diff
@@ -1436,13 +1433,72 @@ PROCEDURE ERRORHANDLER
+** UX does not open a conversation in Banno People.
+*]
+ EMAILSTART(PARAMOVERRIDEEMAIL,PARAMOVERRIDEEMAIL,"Banno Address Change - Intervention Needed",EMAILSENDERROR)
+ EMAILLINE("MIME-Version: 1.0",EMAILSENDERROR)
```
Done, I changed the error detail to show the first 60 characters of any email error encountered. If none, then it says where the email was sent.

jucarson commented on change-address/REPWRITERSPECS/BANNO.CHANGE.ADDR.V1.POW:
```diff
@@ -1584,6 +1642,21 @@ PROCEDURE READCONFIGFILESETTINGS
+[* Override conversations
+*]
+       ELSE IF PARAMNAME="OC" AND
+          PARAMVALUE<>"" THEN
```
Done

jucarson commented on change-address/REPWRITERSPECS/BANNO.CHANGE.ADDR.V1.POW:
```diff
@@ -1436,13 +1433,72 @@ PROCEDURE ERRORHANDLER
+ EMAILLINE("Error Detail : "+ERRORDETAIL(ERRORDETAILOFFSET),EMAILSENDERROR)
+ EMAILLINE("<br /><br />",EMAILSENDERROR)
+ EMAILLINE("New address information can be found in the account notes.",EMAILSENDERROR)
+ EMAILLINE("<br />",EMAILSENDERROR)
```
I tested adding it and it worked with no problem, but then I realized that the locator is not visible anywhere in the account manager, so I did not add anything for this to this commit.

jucarson commented on change-address/LETTERSPECS/BANNO.CHANGE.ADDR.V1.CFG:
```diff
@@ -128,42 +128,62 @@ MT: 0
 SE:
 *
-* EMAIL TYPE: 
+* MEMBER CONIRMATION EMAIL TYPE: 
```
Done

jucarson commented:
Jordan, I have added my responses below in **bold** to make them easy to find.
> 
> 1. The way I read the Error codes, 503 means Error processing requested changes.
>    So in the following if condition statement, shouldn't RESPONSECODE<>504 actually
>    be RESPONSECODE=503?, since I don't think the other error codes would trigger conversations,
>    as they are memo mode, config error and ineligible account type errors? **Currently all result codes besides 504 (success) trigger a conversations and tell the member it will be reviewed.**
>    IF ((PARAMOVERRIDECONVS=TRUE AND
>    RESPONSECODE<>504) AND
>    (PARAMTESTACCTS="" OR
>    (PARAMTESTACCTS<>"" AND
>    TESTUSER=TRUE))) THEN
>    CALL OVERRIDECONVERSATIONS
>    Also, there is an exception to what error code 503 means in one place in this PowerOn's code that I've pasted here;
>    so I don't know if you would need to consider that too (it looks like in that case a change didn't fail, just that there were no names that could be changed, but I'm not sure)?  **It is still an error that would trigger conversations, so we would still want to send the email and notify the CU.**
>    IF TOTALNAMECOUNT=0 THEN
>    DO
>    CALL JSONOPEN
>    CALL PRINTPROGRAMINFO
>    RESPONSECODE=503
>    ERROR503CHR="No allowed name records found"
>    PARSEDTEXT(1)="No allowed name records found"
>    CALL ERRORHANDLER
>    CALL CREATENOTERECORD
>    END
> 2. and if it is supposed to be looking for RESPONSECODE=503,
>    looking at these eleven code sections,
>    wouldn't the note record have already been created before
>    the new code adds lines 3-6 to put in the note?  **The code above that prints no records found is happening only if name matching is turned on and it done before the code below. The first of the 11 sections below is called only if name matching is off and the 10 other sections are only called if the response code is equal to zero, which it couldn't be if it found no names above.** 
>    IF NAMELOC=0 THEN
>    DO
>    RESPONSECODE=503
>    PARSEDTEXT(1)=FORMAT("Banno Chg Addr Err: Name LOC:###[Updating files with current versions #9](https://github.com/Banno/banno-powerons/pull/9)",NAMELOC)
>    ERROR503CHR=FORMAT("Could not locate name LOC:###[Updating files with current versions #9](https://github.com/Banno/banno-powerons/pull/9)",NAMELOC)
>    CALL CREATENOTERECORD
>    END
>    END
>    ...
>    IF FMERROR<>"" THEN
>    DO
>    RESPONSECODE=503
>    ERROR503CHR="Banno Chg Addr Err: FM-"+FMERROR
>    PARSEDTEXT(1)="Banno Chg Addr Err: FM-"+FMERROR
>    CALL CREATENOTERECORD
>    END
>    ...
>    ELSE IF FMERROR<>"" THEN
>    DO
>    RESPONSECODE=503
>    ERROR503CHR=FORMAT("Banno Chg Addr Err - Nm LOC ####[Updating files with current versions #9](https://github.com/Banno/banno-powerons/pull/9)",ACCTLVLLOCS(TMPLOOP))+" FM-"+FMERROR
>    PARSEDTEXT(1)=FORMAT("Banno Chg Addr Err - Nm LOC ####[Updating files with current versions #9](https://github.com/Banno/banno-powerons/pull/9)",ACCTLVLLOCS(TMPLOOP))
>    PARSEDTEXT(2)="FM-"+FMERROR
>    CALL CREATENOTERECORD
>    END
>    ...
>    ELSE IF FMERROR<>"" THEN
>    DO
>    RESPONSECODE=503
>    ERROR503CHR=FORMAT("Banno Chg Addr Err - Shr Nm LOC ####[Updating files with current versions #9](https://github.com/Banno/banno-powerons/pull/9)",SHARELVLLOCS(TMPLOOP))+" FM-"+FMERROR
>    PARSEDTEXT(1)=FORMAT("Banno Chg Addr Err - Shr Nm LOC ####[Updating files with current versions #9](https://github.com/Banno/banno-powerons/pull/9)",SHARELVLLOCS(TMPLOOP))
>    PARSEDTEXT(2)="FM-"+FMERROR
>    CALL CREATENOTERECORD
>    END
>    ...
>    ELSE IF FMERROR<>"" THEN
>    DO
>    RESPONSECODE=503
>    ERROR503CHR=FORMAT("Banno Chg Addr Err - IRS Nm LOC ####[Updating files with current versions #9](https://github.com/Banno/banno-powerons/pull/9)",IRSLVLLOCS(TMPLOOP))+" FM-"+FMERROR
>    PARSEDTEXT(1)=FORMAT("Banno Chg Addr Err - IRS Nm LOC ####[Updating files with current versions #9](https://github.com/Banno/banno-powerons/pull/9)",IRSLVLLOCS(TMPLOOP))
>    PARSEDTEXT(2)="FM-"+FMERROR
>    CALL CREATENOTERECORD
>    END
>    ...
>    ELSE IF FMERROR<>"" THEN
>    DO
>    RESPONSECODE=503
>    ERROR503CHR=FORMAT("Banno Chg Addr Err - Ln Nm LOC ####[Updating files with current versions #9](https://github.com/Banno/banno-powerons/pull/9)",LOANLVLLOCS(TMPLOOP))+" FM-"+FMERROR
>    PARSEDTEXT(1)=FORMAT("Banno Chg Addr Err - Ln Nm LOC ####[Updating files with current versions #9](https://github.com/Banno/banno-powerons/pull/9)",LOANLVLLOCS(TMPLOOP))
>    PARSEDTEXT(2)="FM-"+FMERROR
>    CALL CREATENOTERECORD
>    ...
>    ELSE IF FMERROR<>"" THEN
>    DO
>    RESPONSECODE=503
>    ERROR503CHR=FORMAT("Banno Chg Addr Err - Pldg Nm LOC ####[Updating files with current versions #9](https://github.com/Banno/banno-powerons/pull/9)",PLEDGELVLLOCS(TMPLOOP))+" FM-"+FMERROR
>    PARSEDTEXT(1)=FORMAT("Banno Chg Addr Err - Pldg Nm LOC ####[Updating files with current versions #9](https://github.com/Banno/banno-powerons/pull/9)",PLEDGELVLLOCS(TMPLOOP))
>    PARSEDTEXT(2)="FM-"+FMERROR
>    CALL CREATENOTERECORD
>    END
>    ...
>    ELSE IF FMERROR<>"" THEN
>    DO
>    RESPONSECODE=503
>    ERROR503CHR=FORMAT("Banno Chg Addr Err - EFT Nm LOC ####[Updating files with current versions #9](https://github.com/Banno/banno-powerons/pull/9)",EFTLVLLOCS(TMPLOOP))+" FM-"+FMERROR
>    PARSEDTEXT(1)=FORMAT("Banno Chg Addr Err - EFT Nm LOC ####[Updating files with current versions #9](https://github.com/Banno/banno-powerons/pull/9)",EFTLVLLOCS(TMPLOOP))
>    PARSEDTEXT(2)="FM-"+FMERROR
>    CALL CREATENOTERECORD
>    END
>    ...
>    ELSE IF FMERROR<>"" THEN
>    DO
>    RESPONSECODE=503
>    ERROR503CHR=FORMAT("Banno Chg Addr Err - Card Nm LOC ####[Updating files with current versions #9](https://github.com/Banno/banno-powerons/pull/9)",CARDLVLLOCS(TMPLOOP))+" FM-"+FMERROR
>    PARSEDTEXT(1)=FORMAT("Banno Chg Addr Err - Card Nm LOC ####[Updating files with current versions #9](https://github.com/Banno/banno-powerons/pull/9)",CARDLVLLOCS(TMPLOOP))
>    PARSEDTEXT(2)="FM-"+FMERROR
>    CALL CREATENOTERECORD
>    END
>    ...
>    ELSE IF FMERROR<>"" THEN
>    DO
>    RESPONSECODE=503
>    ERROR503CHR=FORMAT("Banno Chg Addr Err - X Ln Nm LOC ####[Updating files with current versions #9](https://github.com/Banno/banno-powerons/pull/9)",EXTERNALLVLLOCS(TMPLOOP))+" FM-"+FMERROR
>    PARSEDTEXT(1)=FORMAT("Banno Chg Addr Err - X Ln Nm LOC ####[Updating files with current versions #9](https://github.com/Banno/banno-powerons/pull/9)",EXTERNALLVLLOCS(TMPLOOP))
>    PARSEDTEXT(2)="FM-"+FMERROR
>    CALL CREATENOTERECORD
>    END
>    ...
>    IF TRACKINGFOUND=FALSE OR
>    (TRACKINGFOUND=TRUE AND MEMBERADDRLINK="") THEN
>    DO
>    RESPONSECODE=503
>    IF TRACKINGFOUND=FALSE THEN
>    DO
>    PARSEDTEXT(1)="Account tracking type 08 not found"
>    ERROR503CHR="Tracking 8 not found"
>    END
>    ELSE IF (TRACKINGFOUND=TRUE AND MEMBERADDRLINK="") THEN
>    DO
>    PARSEDTEXT(1)="Tracking 8 found but no Mbr Addr link"
>    ERROR503CHR="Tracking 8 without MbrAddr Link"
>    END
>    CALL CREATENOTERECORD
> 3. In the case where an error occurs during address change,
>    I think it would be a good Idea to send an email to the member
>    at the same time as making the note record
>    with similar language that Banno shows members for when an address
>    change has an error and has to go to conversations.
>    This way the member would know there was an issue,
>    that will be manually changed by the CU, and that would
>    give them the idea possibly double check their address
>    in Banno in a few days to make sure the cu made the manual change.
>    You might want to ask the BA's to see if they think it should
>    be part of it.
>    Or another way would add a cu customizable
>    message for that saturation to the cfg and contract that,
>    gets a different new error code,
>    so the UX Team can display the custom message to the member, without sending them to conversations.  **I spoke about this with Emily and Tiffany during development and since the only way for the PowerOn to stop the send to conversations is to send the UX the success code (504). When the UX gets that code, it assumes that Symitar posted it correctly and actually changes that information in the Banno database. So, until the next time a DB refresh is done, the address change looks successful. Since we cannot change the UX as part of the project, they decided that letting it looks successful was the best option and that the CU would most likely fix it (or contact the member) before that DB refresh. All that in mind, an email to the member would be confusing to them as they are assuming it was successful.** 
> 4. I think adding Debug to the Email sending would be good (like is in
>    the procedure SENDEMAIL). That way the logs also would have any email error
>    like this code in the SENDMAIL procedure:
>    BNODEBUGMSG= "SEND EMAIL ERROR:"+EMAILSENDERROR
>    CALL BNODEBUGMSGADD  **Tom also suggested the same thing. I have added error handling to the OVERRIDECONVERSATIONS procedure that changes the logging message to show an email error if there was one. If not, it shows the email it was sent to like it did before this commit.**
> 5. Miscellaneous spelling to look at.
> 
> I know this is old code but Seperate could be changed to Separate here: **Done**
> 
> PROCEDURE PARSEDATALINE [* Parse through a line of data, looking for a pipe delimiter. ** Seperate text at the delimiter storing each segment into ** array elements
> 
> Same here: sucessfull can be changed to successful **Done**
> 
> PROCEDURE READCONFIGFILESETTINGS [* Read parameter settings from the configuration Letter file. ** IAT PARAMINELIGIBLEACCTS Ineligible Account Type(s) (in LISTEXPAND format) ** AW PARAMWARNINGSACCT Acct level warnings which exclude account (in LISTEXPAND format) ** CW PARAMCLEARWARNING Account level warning(s) to clear upon successfull address change
> 
> Same here:tto can be changed to to.     **Done**
> 
> PROCEDURE POSTFMPROCESS [* Clears and sets warning codes upon successful FM. Sends ** confirmation email tto member.

jordan-keenan reviewed (commented):
Thanks you  for explanation, Justin. It helped me understand how it works when ELSE IF PERFORMMATCHING=TRUE. And thank for adding the additional error details and correcting old program spelling errors.

I still seem to see an issue here.
Procedure Findtracking called when PERFORMMATCHING=FALSE, does not call the error handler when it fails, so the while the error handler eventually gets called in JSON,
the noterecord is already done being created before then.
So the new procedure to override conversations though called in RETURNJSON1, won't result in the failed requested address change being added to the note record in that case.
I'm not sure if you can fix this by calling the error handler right before Note creation?
Or some other way like in the section for ELSE IF PERFORMMATCHING=TRUE, which calls 
the Error handler one last time at the end, and creates a note record. Or you could create
your own note record.

Also, right after the call to FINDTRACKING,
If it fails to find the NAME:LOCATOR to change you end up in the same situation.
while the error handler eventually gets called in JSON,
the noterecord is already done being created before then,
So the new procedure to override conversations though called in RETURNJSON1, won't result in the failed requested address change being added to the note record in that case.

     IF PERFORMMATCHING=FALSE THEN
      DO
       CALL FINDTRACKING

       IF RESPONSECODE=0 THEN
        DO
         FOR EACH NAME WITH (NAME:MBRADDRNUMLINK=MEMBERADDRLINK)
          DO
           NAMELOC=NAME:LOCATOR
          END
         IF NAMELOC=0 THEN
          DO
           RESPONSECODE=503
           PARSEDTEXT(1)=FORMAT("Banno Chg Addr Err: Name LOC:####9",NAMELOC)
           ERROR503CHR=FORMAT("Could not locate name LOC:####9",NAMELOC)
           CALL CREATENOTERECORD
          END
        END

Same issue for the rest of the PERFORMMATCHING=FALSE section
       IF NAMELOC>0 THEN
        DO
         FMERRORCOUNT=0

         FMPERFORM REVISE NAME LOC NAMELOC (0,0,FMERROR)
          DO
           SET EXTRAADDRESS TO UPPERCASE(NEWEXTRAADDR)
           SET STREET TO UPPERCASE(NEWSTREET)
           SET CITY TO UPPERCASE(NEWCITY)
           SET STATE TO UPPERCASE(NEWSTATE)
           SET ZIPCODE TO UPPERCASE(NEWZIPCODE)
          END

         BNODEBUGMSG=FORMAT("NAME LOC #####9 TARGETED. ",NAMELOC)+"FMERROR='"+FMERROR+"'"
         CALL BNODEBUGMSGADD

         IF FMERROR<>"" THEN
          DO
           RESPONSECODE=503
           ERROR503CHR="Banno Chg Addr Err: FM-"+FMERROR
           PARSEDTEXT(1)="Banno Chg Addr Err: FM-"+FMERROR
           CALL CREATENOTERECORD
          END
         ELSE
          DO
           ADDRCHGSUCCESSFUL=TRUE
           CALL POSTFMPROCESS
          END [FMERROR=""]
        END [IF NAMELOC>0]
      END [IF PERFORMMATCHING=FALSE]

PROCEDURE FINDTRACKING
[* Find the tracking 8 record which contains the member's Banno ID and
** retrieve the member's address link number.
** PASS    PREFLOC         Preference record locator of member
** RETURN  TRACKINGFOUND   Was tracking record found
** RETURN  MEMBERADDRLINK  Targeted member's address number link
*]
 TMPCHR=FORMAT("#########9",PREFLOC)
 CALL NLTS
 TRACKINGFOUND=FALSE
 MEMBERADDRLINK=""
 FOR EACH TRACKING WITH (TRACKINGFOUND=FALSE AND
                       ((TRACKING:EXPIREDATE=DATENULL OR
                         TRACKING:EXPIREDATE>SYSTEMDATE) AND
                         TRACKING:USERCHAR12=TMPCHR AND
                         UPPERCASE(TRACKING:USERCHAR13)=CURRBANNOID))
  DO
   TRACKINGFOUND=TRUE
   MEMBERADDRLINK=TRACKING:USERCHAR15
  END

 IF TRACKINGFOUND=FALSE OR
   (TRACKINGFOUND=TRUE AND MEMBERADDRLINK="") THEN
  DO
   RESPONSECODE=503

   IF TRACKINGFOUND=FALSE THEN
    DO
     PARSEDTEXT(1)="Account tracking type 08 not found"
     ERROR503CHR="Tracking 8 not found"
    END
   ELSE IF (TRACKINGFOUND=TRUE AND MEMBERADDRLINK="") THEN
    DO
     PARSEDTEXT(1)="Tracking 8 found but no Mbr Addr link"
     ERROR503CHR="Tracking 8 without MbrAddr Link"
    END
   CALL CREATENOTERECORD
	  END
END [PROCEDURE]

If you do end up making changes for the above,
I recommend, the following too (maybe we never had a space at the end of this one, I don't know):
The program is missing a space after #INCLUDE "RB.LISTEXPAND"
I recommend adding it so when CU's self install they
won't get the error for no space at end when installing for on demand use.

jucarson commented:
Jordan,

Thanks for finding that. You are correct. I have updated the PO so the procedure ADDADDRDETAILSTONOTE is called by the procedure CREATENOTERECORD instead of calling it in OVERRIDECONVERSATIONS (within ERRORHANDLER). Since CREATENOTRECORD is only called when there is an error it works seamlessly to correct all the situations you mentioned. 

Tom, I did not ask for a re-review from you, but feel free! Thanks!

tkainz reviewed (approved):
Just one more quick change you may want to consider

jucarson commented on change-address/LETTERSPECS/BANNO.CHANGE.ADDR.V1.CFG:
```diff
@@ -128,42 +128,64 @@ MT: 0
+*  !!! Please ensure this is a valid email address as notifications !!!
+*  !!! of the validation error cannot be sent.                      !!!
+*  VALID VALUES: valid email address 
+*  DEFAULT: None, value must be specified if "OC:" is set to true.
```
Changed true to "YES". Thanks Tom!

jordan-keenan reviewed (commented):
Hi Justin,
 Thanks for fixing that.

I only found one last thing.
PARSEDTEXT() can start with PARSEDTEXT(2) up to PARSEDTEXT(4) before the new procedure overwrites it PARSEDTEXT(3) and PARSEDTEXT(4),
when it is first filed with the member note from @RGUSERCHR3 for up to 110 long.

Can you find a way to fix that as well? It looks like the last block to this enhancement.

Here is the code I'm talking about that can already pre-write from PARSEDTEXT(2) up to PARSEDTEXT(4) before the note is created that
can then be fill PARSEDTEXT(2), PARSEDTEXT(3) and PARSEDTEXT(4) again between 2 to 6 with new address when conversations is off, and a note is created.


[* member note
*]
   DATALINE=SEGMENT(@RGUSERCHR3,1,110)
   IF DATALINE<>"" THEN
    DO
     TMPCHR=DATALINE
     CALL CLEANUPTEXT
     DATALINE=TMPCHR
    END
   DELIMITER="|"
   CALL PARSEDATALINE
   PARSETEXTRAW=DATALINEFIELD(1)
   PARSEDTEXTSTART=2
   PARSETEXTLENGTH=40
   CALL PARSETEXTFIELD
   BNODEBUGMSG="RG3:"+SEGMENT(@RGUSERCHR3,1,110)
   CALL BNODEBUGMSGADD
   BNODEBUGMSG=FORMAT("TEXT LINES=9",PARSEDTEXTCOUNT)
   CALL BNODEBUGMSGADD
   IF PARSEDTEXTCOUNT>0 THEN
    DO
     FOR TMPLOOP=1 TO PARSEDTEXTCOUNT
      DO
       BNODEBUGMSG=FORMAT("NOTE:9",TMPLOOP)+PARSEDTEXT(TMPLOOP)
       CALL BNODEBUGMSGADD
      END
    END

PROCEDURE PARSETEXTFIELD
[* Parse through a single text variable with a max length of 110
** characters to create an array of the text where each element
** of the array is limited to a max number of characters. Procedure will create
** a segment break at the first space found prior to the max segment length. If
** no space is found, the entire segment length will be stored into the array.
**
**   PASS   PARSETEXTLENGTH    Max length of each parsed text segment
**   PASS   PARSEDTEXTSTART    The PARSEDTEXT array element to start storing the
**                             segments in
**   PASS   PARSETEXTRAW       The original text string
**   RETURN PARSEDTEXTCOUNT    The number of array elements of the parsed text
**   RETURN PARSEDTEXT         An array of the parsed text segments
*]
 PARSEDTEXTCOUNT=0
 IF PARSEDTEXTSTART>0 THEN
  PARSEDTEXTCOUNT=PARSEDTEXTSTART-1

 BNODEBUGMSG="PARSETEXTRAW="+PARSETEXTRAW
 CALL BNODEBUGMSGADD

 WHILE PARSETEXTRAW<>""
  DO
   IF LENGTH(PARSETEXTRAW)<=40 THEN
    DO
     IF PARSEDTEXTCOUNT<PARSEDTEXTMAX THEN
      DO
       PARSEDTEXTCOUNT=PARSEDTEXTCOUNT+1
       PARSEDTEXT(PARSEDTEXTCOUNT)=PARSETEXTRAW
       PARSETEXTRAW=""
      END
    END
   ELSE
    DO
     POS=40
     WHILE (FOUND=FALSE AND POS>0)
      DO
       IF SEGMENT(PARSETEXTRAW,POS,POS)=" " THEN
        FOUND=TRUE
       ELSE
        POS=POS-1
      END
     IF FOUND=FALSE THEN
      DO [no space found in first 40 characters - grab first 40 characters]
       IF PARSEDTEXTCOUNT<PARSEDTEXTMAX THEN
        DO
         PARSEDTEXTCOUNT=PARSEDTEXTCOUNT+1
         PARSEDTEXT(PARSEDTEXTCOUNT)=SEGMENT(PARSETEXTRAW,1,40)
         PARSETEXTRAW=SEGMENT(PARSETEXTRAW,41,LENGTH(PARSETEXTRAW))
        END
      END
     ELSE
      DO [space found. grab chars 1 through space-1, leave remaining]
       IF PARSEDTEXTCOUNT<PARSEDTEXTMAX THEN
        DO
         PARSEDTEXTCOUNT=PARSEDTEXTCOUNT+1
         PARSEDTEXT(PARSEDTEXTCOUNT)=SEGMENT(PARSETEXTRAW,1,POS-1)
         PARSETEXTRAW=SEGMENT(PARSETEXTRAW,POS+1,LENGTH(PARSETEXTRAW))
        END
      END
    END
  END

jucarson commented:
Thanks for catching that Jordan. After looking into it further, I noticed that if PERFORMMATCHING is TRUE, it is incrementing the PARSEDTEXTCOUNT if it gets more than 1 error. So, since adding it to the note code can be unpredictable, I have changed the procedure ADDADDRDETAILSTONOTE to ADDADDRDETAILSNOTE which, instead of inserting the address information in the current note, adds it to a separate note altogether. This addresses your concern, and the FMERROR during matching concern. Please re-review when you can. There is a CU anxiously awaiting this update. Thanks!

jordan-keenan reviewed (approved):
The Changes look good! Thanks for making this happen, Justin.

jordan-keenan reviewed (approved):
Changes look good!

tkainz reviewed (approved):
Justin, changes look good - Thanks
